src/framework/myCFG.ml

```diff
@@ -165,138 +165,214 @@ let createCFG (file: file) =
   let cfgB = H.create 113 in
   if Messages.tracing then Messages.trace "cfg" "Starting to build the cfg.\n\n";
 
-  (* Utility function to add stmt edges to the cfg *)
-  let addCfg' t xs f =
+  let addEdges fromNode edges toNode =
     if Messages.tracing then
       Messages.trace "cfg" "Adding edges [%a] from\n\t%a\nto\n\t%a ... "
-        pretty_edges xs
-        pretty_short_node f
-        pretty_short_node t;
-    H.add cfgB t (xs,f);
-    H.add cfgF f (xs,t);
+        pretty_edges edges
+        pretty_short_node fromNode
+        pretty_short_node toNode;
+    H.add cfgB toNode (edges,fromNode);
+    H.add cfgF fromNode (edges,toNode);
     Messages.trace "cfg" "done\n\n"
   in
-  let addCfg t (e,f) = addCfg' t [getLoc f,e] f in
-  let mkEdge fromNode edge toNode = addCfg (Statement toNode) (edge, Statement fromNode) in
-  let mkEdges fromNode edges toNode = addCfg' toNode edges fromNode in
-  (* Function for finding the next real successor of a statement. CIL tends to
-   * put a lot of junk between stuff: *)
-  let realnode is_entry stmt =
-    let rec realnode is_entry visited stmt =
-      if List.mem stmt.sid visited then stmt
-      else match (List.hd stmt.succs) with
-        | exception (Failure _) -> if is_entry then stmt else raise Not_found
-        | next -> begin
-            let sid = stmt.sid in
-            match stmt.skind with
-            | Block _ -> realnode is_entry (sid::visited) next
-            | Goto _ -> realnode is_entry (sid::visited) next
-            | Instr [] -> begin
-                if next.sid == stmt.sid
-                then stmt
-                else realnode is_entry (sid::visited) next
-              end
-            | Loop _ -> realnode is_entry (sid::visited) next
-            | If (exp,_,_,_) -> if isZero exp then realnode is_entry (sid::visited) next else stmt
-            | _ -> stmt
+  let addEdge fromNode edge toNode = addEdges fromNode [edge] toNode in
+  let addEdge_fromLoc fromNode edge toNode = addEdge fromNode (getLoc fromNode, edge) toNode in
```

Comment:
These two things both called `edge` are of different types, right? Maybe we should make this clear here somehow?

Reply:
Indeed they are slightly different things. We use `MyCFG.edge` variants for the CFG edge labels, that's what most of Goblint sees. But inside the CFG modules, they always are paired with locations. Since this is the CFG construction, all of this code considers edges in the latter sense.

I would have wanted to even get rid of `addEdge_fromLoc` but there are a few special cases here, where the desired location doesn't come directly from some pattern matching but must be looked up specially.